src/generators/markdowndocgenerator/index.ts

```diff
@@ -0,0 +1,252 @@
+
+      // First test for '#'-style headers.
+      let match = line.match(/^([#]{1,6}) (.*)/);
+      if (match) {
+        const level = match[1].length;
+        // Take special care to parse markdown links into raw text.
+        const text = match[2].replace(/\[(.+)\]\(.+\)/g, '$1');
+        let numbering = generateNumbering(numberingDict, level);
+        headings.push({
+          text,
+          numbering,
+          level,
+          onClick,
+          type: INotebookHeadingTypes.header
+        });
+        return;
+      }
+
+      // Next test for '==='-style headers.
+      match = line.match(/^([=]{2,}|[-]{2,})/);
+      if (match && idx > 0) {
+        const level = match[1][0] === '=' ? 1 : 2;
+        // Take special care to parse markdown links into raw text.
+        const text = lines[idx - 1].replace(/\[(.+)\]\(.+\)/g, '$1');
+        let numbering = generateNumbering(numberingDict, level);
+        headings.push({
+          text,
+          numbering,
+          level,
+          onClick,
+          type: INotebookHeadingTypes.header
+        });
+        return;
+      }
+
+      // Finally test for HTML headers. This will not catch multiline
+      // headers, nor will it catch multiple headers on the same line.
+      // It should do a decent job of catching many, though.
+      match = line.match(/<h([1-6])>(.*)<\/h\1>/i);
+      if (match) {
+        const level = parseInt(match[1], 10);
+        const text = match[2];
+        let numbering = generateNumbering(numberingDict, level);
+        headings.push({
+          text,
+          numbering,
+          level,
+          onClick,
+          type: INotebookHeadingTypes.header
+        });
+        return;
+      }
+    });
+    return headings;
+  }
+}
+
+namespace Private {
+  /**
+   * Given a HTML DOM element, get the markdown headings
+   * in that string.
+   */
+  export function getRenderedHTMLHeadingsForMarkdownDoc(
+    node: HTMLElement,
+    onClickFactory: (el: Element) => (() => void),
+    sanitizer: ISanitizer,
+    numberingDict: { [level: number]: number },
+    needNumbering = true
+  ): INotebookHeading[] {
+    let headings: INotebookHeading[] = [];
+    let headingNodes = node.querySelectorAll('h1, h2, h3, h4, h5, h6');
+    for (let i = 0; i < headingNodes.length; i++) {
+      const heading = headingNodes[i];
+      const level = parseInt(heading.tagName[1]);
+      let text = heading.textContent ? heading.textContent : '';
+      let shallHide = !needNumbering;
+
+      // Show/hide numbering DOM element based on user settings
+      if (heading.getElementsByClassName('numbering-entry').length > 0) {
+        heading.removeChild(
+          heading.getElementsByClassName('numbering-entry')[0]
+        );
+      }
+      let html = sanitizer.sanitize(heading.innerHTML, sanitizerOptions);
+      html = html.replace('¶', ''); // Remove the anchor symbol.
+      const onClick = onClickFactory(heading);
```

Comment:
I think the `onClickFactory` is less necessary now that the rendered markdown generator and the notebook generators share less code when parsing them for header elements. Perhaps we can simplify the logic by removing that from the function signature?

Reply:
I already removed the onClickFactory for ```getRenderedHTMLHeadingsForMarkdownDoc``` function. 

But for the ```getMarkdownDocHeadings``` function, its onClick needs ```editor```, which is an argument of generator, and we have to pass it into  ```getMarkdownDocHeadings``` if we remove onClickFactory. I was wondering is it worth it for us to remove one argument but add another one to it?